File: leetcode/src/medium/PathSumII.java
package medium;

import java.util.ArrayList;
import java.util.List;

public class PathSumII {
	List <List <Integer>> ans;
	int sum;
	
	public List<List<Integer>> pathSum(TreeNode root, int sum) {
		this.ans = new ArrayList <>();
		this.sum = sum;
		
		List <Integer> lst = new ArrayList <>();
		recurse(root, 0, 0, lst);
		return ans;
	}

	private void recurse(TreeNode root, int runningSum, int height, List<Integer> lst) {
		if (root == null) return;
		
		if (lst.size() > height) {
			lst.set(height, root.val);
		} else {
			lst.add(root.val);
		}
		
		if (root.left == null && root.right == null) {
			
			if (runningSum + root.val == sum) {
				List <Integer> kk = new ArrayList <>();
				for (int k = 0; k < height + 1; k++) {
					kk.add(lst.get(k));
				}
				ans.add(kk);
			}
		}
		
		recurse(root.left, runningSum + root.val, height + 1, lst);
		recurse(root.right, runningSum + root.val, height + 1, lst);
	}
	
	public static void main(String[] args) {
		PathSumII sol = new PathSumII();
		
		TreeNode root = new TreeNode(5);
		root.left = new TreeNode(4);
		root.right = new TreeNode(8);
		root.left.left = new TreeNode(11);
		root.right.left = new TreeNode(13);
		root.right.right = new TreeNode(4);
		root.left.left.left = new TreeNode(7);
		root.left.left.right = new TreeNode(2);
		root.right.right.left = new TreeNode(5);
		root.right.right.right = new TreeNode(1);
   		
		sol.pathSum(root, 22);
		System.out.println(sol.ans);
	}
}

File: leetcode-cpp/archieve/PathSumII.cpp
#include "header.h"

void PathSumII::recurse(TreeNode* root, int running_sum, int height, vector <int> v) {
	
	if (root == NULL) return;
	if (v.size() > height) {
		v.insert(v.begin() + height, root->val);
	} else {
		v.push_back(root->val);
	}

	if (root->left == NULL && root->right == NULL) {
		if (running_sum + root->val == sum) {
			vector <int> kk;
			for (int k = 0; k < height + 1; k++)
				kk.push_back(v[k]);
			ans.push_back(kk);
		}
	}
	recurse(root->left, running_sum + root->val, height + 1, v);
	recurse(root->right, running_sum + root->val, height + 1, v);
}

vector<vector<int>> PathSumII::pathSum(TreeNode *root, int sum) {
	this->sum = sum;
	vector <int> v;
	recurse(root, 0, 0, v);
	return ans;
}

